--!strict
--!native

local DelaunayTriangle = require("./InternalClasses/DelaunayTriangle")
local DelaunayTriangleEdge = require("./InternalClasses/DelaunayTriangleEdge")
local DelaunayTriangleSet = require("./InternalClasses/DelaunayTriangleSet")
local PointBinGrid = require("./InternalClasses/PointBinGrid")

local DelaunayUtil = require("./Utils/DelaunayUtil")
local Triangle2dUtil = require("./Utils/Triangle2dUtil")
local DrawUtil = require("./Utils/DrawUtil")

export type DelaunayTriangle = DelaunayTriangle.DelaunayTriangle
export type DelaunayTriangleEdge = DelaunayTriangleEdge.DelaunayTriangleEdge
export type DelaunayTriangleSet = DelaunayTriangleSet.DelaunayTriangleSet
type PointBinGrid = PointBinGrid.PointBinGrid
type Triangle2d = Triangle2dUtil.Triangle2d

type bounds = {
	min: Vector2,
	max: Vector2,
	center: Vector2,
	size: Vector2,
}

type CDT = {
	TriangleSet: DelaunayTriangleSet,
	DiscardedTriangles: { number },
	Grid: PointBinGrid?,
	AdjacentTriangleStack: { number },
	AdjacentTriangleEdgeStack: { number },

	GetTrianglesDiscardingHoles: (self: CDT, outputTriangles: { Triangle2d }?) -> { Triangle2d },
	GetAllTriangles: (self: CDT, _outputTriangles: { Triangle2d }?) -> { Triangle2d },
	Render: (self: CDT) -> (),
	Triangulate: (
		self: CDT,
		inputPoints: { Vector2 },
		maximumAreaTesselation: number?,
		polygonHoles: { { Vector2 } }?
	) -> (),

	GetOutlinePolygon: (self: CDT) -> { Vector2 },
	GetSharedEdge: (self: CDT, triangle: DelaunayTriangle, adjacentTriangleIndex: number) -> number,
	GetSupertriangleTriangles: (self: CDT, _outputTriangles: { number }?) -> { number },
	CalculateBoundsWithLeftBottomCornerAtOrigin: (self: CDT, points: { Vector2 }) -> bounds,
	NormalizePoints: (self: CDT, points: { Vector2 }, bounds: bounds) -> { Vector2 },
	DenormalizePoints: (self: CDT, points: { Vector2 }, bounds: bounds) -> { Vector2 },
	Tesselate: (self: CDT, maximumTriangleArea: number) -> (),
}
export type ConstrainedDelaunayTriangulation = CDT

type CDTInternal = {
	_PointCloudBounds: bounds?,
	_CachedTriangles: { Triangle2d }?,

	_AddPointToTriangulation: (self: CDTInternal, pointToInsert: Vector2) -> number,
	_FulfillDelaunayConstraint: (
		self: CDTInternal,
		adjacentTrianglesToProcess: { number },
		adjacentTriangleEdges: { number }
	) -> (),
	_SwapEdges: (
		self: CDTInternal,
		mainTriangleIndex: number,
		mainTriangle: DelaunayTriangle,
		notInEdgeVertexLocalIndex: number,
		oppositeTriangle: DelaunayTriangle,
		oppositeTriangleSharedEdgeVertexLocalIndex: number
	) -> (),
	_AddConstrainedEdgeToTriangulation: (
		self: CDTInternal,
		endpointAIndex: number,
		endpointBIndex: number
	) -> (),
} & CDT

-- Toggle on to make debug rendering easier
local DISABLE_NORMALIZATION = false

local NOT_FOUND = -1
local NO_ADJACENT_TRIANGLE = -1

local profilebegin = debug.profilebegin
local profileend = debug.profileend
local IsPointInCircumcircle = Triangle2dUtil.IsPointInCircumcircle

--------------------------------------------------------------------------------
--// Class //--
--------------------------------------------------------------------------------

--[=[
    @class ConstrainedDelaunayTriangulation
    Encapsulates the entire constrained Delaunay triangulation algorithm, based on S. W. Sloan's proposal.
    Call `Triangulate` to obtain the triangulation of a point cloud.
]=]
local ConstrainedDelaunayTriangulation = {}
ConstrainedDelaunayTriangulation.__index = ConstrainedDelaunayTriangulation

function ConstrainedDelaunayTriangulation.new(): CDT
	local self = setmetatable({}, ConstrainedDelaunayTriangulation) :: any
	self.TriangleSet = DelaunayTriangleSet.new()
	self.DiscardedTriangles = {}
	self.Grid = nil
	self.AdjacentTriangleStack = {}
	self.AdjacentTriangleEdgeStack = {}
	return self :: CDT
end

--[=[
    Reads the triangles generated by the Triangulate method, discarding all those triangles that are inside a hole or belong to the supertriangle.
    @param outputTriangles {Triangle2d} The list to which the triangles will be added. No elements will be removed from this list.
]=]
function ConstrainedDelaunayTriangulation.GetTrianglesDiscardingHoles(
	self: CDTInternal,
	outputTriangles: { Triangle2d }?
): { Triangle2d }
	local out = outputTriangles or {}

	for i = 1, self.TriangleSet:GetTriangleCount() do
		local isTriangleToBeRemoved = false

		for _, triangleIndex in ipairs(self.DiscardedTriangles) do
			if triangleIndex >= i then
				isTriangleToBeRemoved = triangleIndex == i
				break
			end
		end

		if not isTriangleToBeRemoved then
			local triangle = self.TriangleSet:GetDelaunayTriangle(i)
			table.insert(
				out,
				Triangle2dUtil.new(
					self.TriangleSet.Points[triangle.Points[1]],
					self.TriangleSet.Points[triangle.Points[2]],
					self.TriangleSet.Points[triangle.Points[3]]
				)
			)
		end
	end

	return out
end

--[=[
    Reads all the triangles generated by the Triangulate method, without discarding any.
    @param outputTriangles {Triangle2d} The list to which the triangles will be added. No elements will be removed from this list.
]=]
function ConstrainedDelaunayTriangulation.GetAllTriangles(
	self: CDTInternal,
	_outputTriangles: { Triangle2d }?
): { Triangle2d }
	local outputTriangles = _outputTriangles or {}
	for i = 1, self.TriangleSet:GetTriangleCount() do
		local triangle = self.TriangleSet:GetDelaunayTriangle(i)
		table.insert(
			outputTriangles,
			Triangle2dUtil.new(
				self.TriangleSet.Points[triangle.Points[1]],
				self.TriangleSet.Points[triangle.Points[2]],
				self.TriangleSet.Points[triangle.Points[3]]
			)
		)
	end
	return outputTriangles
end

--[=[
    
]=]
function ConstrainedDelaunayTriangulation.Render(self: CDTInternal)
	if not self._PointCloudBounds then
		return
	end
	self.TriangleSet.Points = self:DenormalizePoints(self.TriangleSet.Points, self._PointCloudBounds)
	local triangles = self:GetAllTriangles()
	for _, triangle in triangles do
		DrawUtil.DrawTriangle(triangle :: { Vector2 | Vector3 }, BrickColor.random().Color)
	end
	self.TriangleSet.Points = self:NormalizePoints(self.TriangleSet.Points, self._PointCloudBounds)
end

--[=[
    Generates the triangulation of a point cloud that fulfills the Delaunay constraint. It allows the creation of holes in the triangulation, formed by closed polygons that do not overlap each other.

    @param inputPoints {Vector2} -- The main point cloud. It must contain at least 3 points.
    @param maximumAreaTesselation number? -- Optional. When greater than zero, all triangles will be tessellated until none exceeds this area.
    @param polygonHoles {{Vector2}}? -- Optional. List of holes defined by closed polygons sorted counter-clockwise.
]=]
function ConstrainedDelaunayTriangulation.Triangulate(
	self: CDTInternal,
	inputPoints: { Vector2 },
	maximumAreaTesselation: number?,
	polygonHoles: { { Vector2 } }?
)
	profilebegin("Triangulate")
	local TriangleSet: DelaunayTriangleSet = self.TriangleSet

	-- Initialize containers
	TriangleSet:clear()
	table.clear(self.AdjacentTriangleStack)
	table.clear(self.AdjacentTriangleEdgeStack)
	table.clear(self.DiscardedTriangles)

	-- 1: Normalization
	local mainPointCloudBounds = self:CalculateBoundsWithLeftBottomCornerAtOrigin(inputPoints)
	self._PointCloudBounds = mainPointCloudBounds
	local normalizedPoints = self:NormalizePoints(inputPoints, mainPointCloudBounds)

	-- 2: Addition of points to the space partitioning grid
	local gridSize = math.ceil(math.sqrt(math.sqrt(#inputPoints)))
	local normalizedCloudBounds = self:CalculateBoundsWithLeftBottomCornerAtOrigin(normalizedPoints)
	self.Grid = PointBinGrid.new(gridSize, normalizedCloudBounds.size)
	local grid = self.Grid
	if not grid then
		profileend()
		return
	end

	for _, point in normalizedPoints do
		grid:AddPoint(point)
	end

	-- 3: Supertriangle initialization
	-- local superSize = 100
	-- local supertriangle = {Vector2.new(-superSize, -superSize), Vector2.new(superSize, -superSize), Vector2.new(0, superSize)}
	local supertriangle = DelaunayUtil.CreateSuperTriangle(inputPoints)

	TriangleSet:AddRawTriangle(supertriangle[1], supertriangle[2], supertriangle[3], -1, -1, -1)

	-- 4: Adding points to the Triangle set and Triangulation
	-- Points are added one at a time, and points that are close together are inserted together because they are
	-- sorted in the grid, so a later step for finding their containing triangle is faster
	-- print("4: Adding points to the triangulation")
	grid:ForEachPoint(function(point)
		self:_AddPointToTriangulation(point)
	end)

	if maximumAreaTesselation and maximumAreaTesselation > 0 then
		self:Tesselate(maximumAreaTesselation)
	end

	-- 5: Holes creation (constrained edges)
	if polygonHoles and #polygonHoles > 0 then
		profilebegin("Handle polygon holes")
		-- print("5: Adding constrained edges to the triangulation")
		local constrainedEdgeIndices: { { number } } = {}

		-- Adds the points of all the polygons to the triangulation
		for _, polygon in polygonHoles do
			profilebegin("Initialize polygon")
			-- 5.1: Normalize
			local normalizedPolygon = self:NormalizePoints(polygon, mainPointCloudBounds)
			local polygonEdgeIndices: { number } = {}

			-- 5.2: Add points to the triangle set
			for _, point in normalizedPolygon do
				-- TODO: check for zero length edge
				local addedPointIndex = self:_AddPointToTriangulation(point)
				table.insert(polygonEdgeIndices, addedPointIndex)
			end

			table.insert(constrainedEdgeIndices, polygonEdgeIndices)
			profileend()
		end

		-- 5.3: Create the constrained edges
		for _, polygonIndices in constrainedEdgeIndices do
			for i = 1, #polygonIndices do
				profilebegin("AddConstrainedEdgeToTriangulation")
				local a = polygonIndices[i]
				local b = polygonIndices[i % #polygonIndices + 1]
				self:_AddConstrainedEdgeToTriangulation(a, b)
				profileend()
			end
		end

		-- 5.4 Identify the triangles that are inside the holes
		for _, polygonIndices in constrainedEdgeIndices do
			TriangleSet:GetTrianglesInPolygon(polygonIndices, self.DiscardedTriangles)
		end

		profileend()
		-- Remove all the triangles left that are not part of the main cloud
		-- TODO: How?
	end

	-- 6: Supertriangle removal
	self:GetSupertriangleTriangles(self.DiscardedTriangles)
	table.sort(self.DiscardedTriangles)

	-- 7: Denormalization
	TriangleSet.Points = self:DenormalizePoints(TriangleSet.Points, mainPointCloudBounds)
	profileend()
end

--[=[
    Adds a point to the triangulation, splitting a triangle into 3 pieces and ensuring all triangles fulfill the Delaunay constraint.
    @param pointToInsert Vector2 The point to add to the triangulation.
    @return number The index of the new point in the triangle set.
]=]
function ConstrainedDelaunayTriangulation._AddPointToTriangulation(self: CDTInternal, pointToInsert: Vector2): number
	profilebegin("AddPointToTriangulation")
	-- Note: Adjacent triangle, opposite to the inserted point, is always at index 1
	-- Note 2: Adjacent triangles are stored CCW automatically, their index matches the index of the first vertex in every edge, and it is known that vertices are stored CCW
	local TriangleSet: DelaunayTriangleSet = self.TriangleSet

	-- 4.1: Check point existence
	local existingPointIndex = TriangleSet:GetIndexOfPoint(pointToInsert)
	if existingPointIndex ~= NOT_FOUND then
		return existingPointIndex
	end

	-- 4.2: Search containing triangle
	local containingTriangleIndex = TriangleSet:FindTriangleThatContainsPoint(pointToInsert)
	-- assert(containingTriangleIndex ~= NOT_FOUND, "Point not found in any triangle!")
	local containingTriangle = TriangleSet:GetDelaunayTriangle(containingTriangleIndex)
	-- print("Containing triangle", containingTriangle, "for point", pointToInsert)

	-- 4.3: Store the point
	-- Inserting a new point into a triangle splits it into 3 pieces, 3 new triangles
	local insertedPoint = TriangleSet:AddPoint(pointToInsert)

	-- 4.4: Create 2 new triangles. (We use the existing triangle as our third and just adjust it)
	profilebegin("Create triangles")
	local newTriangle1 = DelaunayTriangle.new(insertedPoint, containingTriangle.Points[1], containingTriangle.Points[2])
	newTriangle1.AdjacentTriangles[1] = NO_ADJACENT_TRIANGLE
	newTriangle1.AdjacentTriangles[2] = containingTriangle.AdjacentTriangles[1]
	newTriangle1.AdjacentTriangles[3] = containingTriangleIndex
	local triangle1Index = TriangleSet:AddTriangle(newTriangle1)

	local newTriangle2 = DelaunayTriangle.new(insertedPoint, containingTriangle.Points[3], containingTriangle.Points[1])
	newTriangle2.AdjacentTriangles[1] = containingTriangleIndex
	newTriangle2.AdjacentTriangles[2] = containingTriangle.AdjacentTriangles[3]
	newTriangle2.AdjacentTriangles[3] = NO_ADJACENT_TRIANGLE
	local triangle2Index = TriangleSet:AddTriangle(newTriangle2)

	-- Sets adjacency between the 2 new triangles
	newTriangle1.AdjacentTriangles[1] = triangle2Index
	newTriangle2.AdjacentTriangles[3] = triangle1Index
	TriangleSet:SetTriangleAdjacency(triangle1Index, newTriangle1.AdjacentTriangles)
	TriangleSet:SetTriangleAdjacency(triangle2Index, newTriangle2.AdjacentTriangles)

	-- Sets the adjacency of the triangles that were adjacent to the original containing triangle
	if newTriangle1.AdjacentTriangles[2] ~= NO_ADJACENT_TRIANGLE then
		TriangleSet:ReplaceAdjacent(newTriangle1.AdjacentTriangles[2], containingTriangleIndex, triangle1Index)
	end

	if newTriangle2.AdjacentTriangles[2] ~= NO_ADJACENT_TRIANGLE then
		TriangleSet:ReplaceAdjacent(newTriangle2.AdjacentTriangles[2], containingTriangleIndex, triangle2Index)
	end
	profileend()

	-- 4.5: Transform containing triangle into the third
	-- Original triangle is transformed into the third triangle after the point has split the containing triangle into 3
	containingTriangle.Points[1] = insertedPoint
	containingTriangle.AdjacentTriangles[1] = triangle1Index
	containingTriangle.AdjacentTriangles[3] = triangle2Index
	TriangleSet:ReplaceTriangle(containingTriangleIndex, containingTriangle)

	-- 4.6: Add new triangles to a stack
	-- Triangles that contain the inserted point are added to the stack for them to be processed by the Delaunay swapping algorithm
	if containingTriangle.AdjacentTriangles[2] ~= NO_ADJACENT_TRIANGLE then
		table.insert(self.AdjacentTriangleStack, containingTriangleIndex)
		table.insert(self.AdjacentTriangleEdgeStack, 2)
	end

	if newTriangle1.AdjacentTriangles[2] ~= NO_ADJACENT_TRIANGLE then
		table.insert(self.AdjacentTriangleStack, triangle1Index)
		table.insert(self.AdjacentTriangleEdgeStack, 2)
	end

	if newTriangle2.AdjacentTriangles[2] ~= NO_ADJACENT_TRIANGLE then
		table.insert(self.AdjacentTriangleStack, triangle2Index)
		table.insert(self.AdjacentTriangleEdgeStack, 2)
	end

	self:_FulfillDelaunayConstraint(self.AdjacentTriangleStack, self.AdjacentTriangleEdgeStack)
	profileend()
	return insertedPoint
end

--[=[
    Process a stack of triangles to ensure they fulfill the Delaunay constraint, swapping edges if necessary.
    @param adjacentTrianglesToProcess {number} The stack of triangles to process.
    @param adjacentTriangleEdges {number} The stack of shared edge indices. (1 to 3)
]=]
function ConstrainedDelaunayTriangulation._FulfillDelaunayConstraint(
	self: CDTInternal,
	adjacentTrianglesToProcess: { number },
	adjacentTriangleEdges: { number }
)
	profilebegin("FulfillDelaunayConstraint")
	-- warn("Fulfill Delaunay Constraint", adjacentTrianglesToProcess, adjacentTriangleEdges)
	local TriangleSet: DelaunayTriangleSet = self.TriangleSet
	local Points = TriangleSet.Points

	while #adjacentTrianglesToProcess > 0 do
		profilebegin("Process triangle")
		local CURRENT_TRIANGLE_INDEX = table.remove(adjacentTrianglesToProcess) :: number -- TODO: Make these table removals more efficient
		local OPPOSITE_TRIANGLE_EDGE_INDEX = table.remove(adjacentTriangleEdges) :: number

		local triangle = TriangleSet:GetDelaunayTriangle(CURRENT_TRIANGLE_INDEX)
		local OPPOSITE_TRIANGLE_INDEX = triangle.AdjacentTriangles[OPPOSITE_TRIANGLE_EDGE_INDEX]

		if OPPOSITE_TRIANGLE_INDEX == NOT_FOUND then
			profileend()
			continue
		end

		local NOT_IN_EDGE_VERTEX_INDEX = if OPPOSITE_TRIANGLE_EDGE_INDEX + 2 > 3
			then OPPOSITE_TRIANGLE_EDGE_INDEX - 1
			else OPPOSITE_TRIANGLE_EDGE_INDEX + 2

		local triangleVertexNotInEdge = Points[triangle.Points[NOT_IN_EDGE_VERTEX_INDEX]]

		if IsPointInCircumcircle(triangleVertexNotInEdge, TriangleSet:GetTrianglePoints(OPPOSITE_TRIANGLE_INDEX)) then
			profilebegin("Add to Stacks")

			local oppositeTriangle = TriangleSet:GetDelaunayTriangle(OPPOSITE_TRIANGLE_INDEX)

			-- Finds the edge of the opposite triangle that is shared with the other triangle, this edge will be swapped
			local sharedEdgeVertexLocalIndex = self:GetSharedEdge(oppositeTriangle, CURRENT_TRIANGLE_INDEX)

			-- Helper function to add triangles to the stack
			local function addToStack(triangleIndex, edgeIndex)
				if
					triangleIndex ~= NO_ADJACENT_TRIANGLE and not table.find(adjacentTrianglesToProcess, triangleIndex)
				then
					table.insert(adjacentTrianglesToProcess, triangleIndex)
					table.insert(adjacentTriangleEdges, edgeIndex)
				end
			end

			-- Add adjacent triangles of the opposite triangle
			local oppositeNextEdgeIndex = if sharedEdgeVertexLocalIndex == 3 then 1 else sharedEdgeVertexLocalIndex + 1
			addToStack(oppositeTriangle.AdjacentTriangles[sharedEdgeVertexLocalIndex], sharedEdgeVertexLocalIndex)
			addToStack(oppositeTriangle.AdjacentTriangles[oppositeNextEdgeIndex], oppositeNextEdgeIndex)

			-- Add adjacent triangles of the current triangle
			local nextEdgeIndex = if NOT_IN_EDGE_VERTEX_INDEX == 3 then 1 else NOT_IN_EDGE_VERTEX_INDEX + 1
			addToStack(triangle.AdjacentTriangles[NOT_IN_EDGE_VERTEX_INDEX], NOT_IN_EDGE_VERTEX_INDEX)
			addToStack(triangle.AdjacentTriangles[nextEdgeIndex], nextEdgeIndex)

			profileend()

			-- Swap edges
			self:_SwapEdges(
				CURRENT_TRIANGLE_INDEX,
				triangle,
				NOT_IN_EDGE_VERTEX_INDEX,
				oppositeTriangle,
				sharedEdgeVertexLocalIndex
			)
		end
		profileend()
	end
	profileend()
end

--[=[
    Given 2 adjacent triangles, it replaces the shared edge with a new edge that joins both opposite vertices. For example, triangles ABC-CBD would become ADC-ABD.

    For the main triangle, its shared edge vertex is moved so the new shared edge vertex is 1 position behind / or 2 forward (if it was 1, now the shared edge is 0).

    @param mainTriangleIndex -- The index of the main triangle.
    @param mainTriangle -- Data about the main triangle.
    @param notInEdgeVertexLocalIndex -- The local index of the vertex not in the shared edge, in the main triangle.
    @param oppositeTriangle -- Data about the triangle that opposes the main triangle.
    @param oppositeTriangleSharedEdgeVertexLocalIndex -- The local index of the vertex where the shared edge begins, in the opposite triangle.
]=]
function ConstrainedDelaunayTriangulation._SwapEdges(
	self: CDTInternal,
	mainTriangleIndex: number,
	mainTriangle: DelaunayTriangle,
	notInEdgeVertexLocalIndex: number,
	oppositeTriangle: DelaunayTriangle,
	oppositeTriangleSharedEdgeVertexLocalIndex: number
)
	profilebegin("SwapEdges")
	-- point index opposite the shared edge
	local oppositeVertex = (oppositeTriangleSharedEdgeVertexLocalIndex + 1) % 3 + 1

	--           3 _|_ a
	--       A3 _   |   _
	--       _      |      _
	--   1 _     A2 |         _  c (opposite vertex)
	--       _      |      _
	--          _   |   _
	--       A1   _ |_
	--              |
	--            2    b

	--           3 _|_
	--       A3 _       _ A2
	--       _             _
	--   1 _________A1_______ 2
	--   a   _             _  c
	--          _       _
	--             _ _
	--              | b
	--

	-- Update vertices and adjacency
	local oppositeTriangleIndex = mainTriangle.AdjacentTriangles[notInEdgeVertexLocalIndex % 3 + 1]
	-- print("Main:", mainTriangleIndex, mainTriangle, "| Opposite:", oppositeTriangleIndex, oppositeTriangle)
	local nextNotInEdgeVertexLocalIndex = notInEdgeVertexLocalIndex % 3 + 1
	mainTriangle.Points[nextNotInEdgeVertexLocalIndex] = oppositeTriangle.Points[oppositeVertex]
	oppositeTriangle.Points[oppositeTriangleSharedEdgeVertexLocalIndex] = mainTriangle.Points[notInEdgeVertexLocalIndex]
	oppositeTriangle.AdjacentTriangles[oppositeTriangleSharedEdgeVertexLocalIndex] =
		mainTriangle.AdjacentTriangles[notInEdgeVertexLocalIndex]
	mainTriangle.AdjacentTriangles[notInEdgeVertexLocalIndex] = oppositeTriangleIndex
	mainTriangle.AdjacentTriangles[nextNotInEdgeVertexLocalIndex] = oppositeTriangle.AdjacentTriangles[oppositeVertex]
	oppositeTriangle.AdjacentTriangles[oppositeVertex] = mainTriangleIndex

	-- assert(oppositeVertex > 0 and oppositeTriangleSharedEdgeVertexLocalIndex ~= oppositeVertex, "Bad opposite vertex! "..oppositeTriangleSharedEdgeVertexLocalIndex..", "..oppositeVertex)
	-- assert(mainTriangleIndex > 0, "Bad main triangle!")
	-- assert(oppositeTriangleIndex > 0, "Bad opposite triangle!")

	-- Replace triangles in the set
	self.TriangleSet:ReplaceTriangle(mainTriangleIndex, mainTriangle)
	self.TriangleSet:ReplaceTriangle(oppositeTriangleIndex, oppositeTriangle)

	-- Update adjacency for affected triangles
	if mainTriangle.AdjacentTriangles[nextNotInEdgeVertexLocalIndex] ~= NO_ADJACENT_TRIANGLE then
		self.TriangleSet:ReplaceAdjacent(
			mainTriangle.AdjacentTriangles[nextNotInEdgeVertexLocalIndex],
			oppositeTriangleIndex,
			mainTriangleIndex
		)
	end
	if oppositeTriangle.AdjacentTriangles[oppositeTriangleSharedEdgeVertexLocalIndex] ~= NO_ADJACENT_TRIANGLE then
		self.TriangleSet:ReplaceAdjacent(
			oppositeTriangle.AdjacentTriangles[oppositeTriangleSharedEdgeVertexLocalIndex],
			mainTriangleIndex,
			oppositeTriangleIndex
		)
	end
	profileend()
end

--[=[
    Adds a constrained edge to the triangulation, ensuring it remains even if it forms triangles that do not fulfill the Delaunay constraint.
    @param endpointAIndex -- The index of the first vertex of the edge.
    @param endpointBIndex -- The index of the second vertex of the edge.
]=]
function ConstrainedDelaunayTriangulation._AddConstrainedEdgeToTriangulation(
	self: CDTInternal,
	endpointAIndex: number,
	endpointBIndex: number
)
	local TriangleSet: DelaunayTriangleSet = self.TriangleSet
	local Points = TriangleSet.Points

	-- Check if the edge already exists
	if TriangleSet:FindTriangleThatContainsEdge(endpointAIndex, endpointBIndex).TriangleIndex ~= NOT_FOUND then
		-- profileend()
		return
	end

	local edgeEndpointA = Points[endpointAIndex]
	local edgeEndpointB = Points[endpointBIndex]

	-- 5.3.1: Search for the triangle that contains the beginning of the new edge
	local triangleContainingA = TriangleSet:FindTriangleThatContainsLineEndpoint(endpointAIndex, endpointBIndex)
	if triangleContainingA == NOT_FOUND then
		-- profileend()
		self:Render()
		DrawUtil.DrawRay(edgeEndpointA, edgeEndpointB - edgeEndpointA, Color3.new(1, 0, 0), 1)
		local p = DrawUtil.DrawPoint(edgeEndpointA, Color3.new(1, 0, 0), 5)
		warn(edgeEndpointA, edgeEndpointB, "|", endpointAIndex, endpointBIndex, p)
		error("Failed to find triangle containing edge endpoint A")
	end

	-- 5.3.2: Get all the triangle edges intersected by the constrained edge
	local intersectedEdges: { DelaunayTriangleEdge } =
		TriangleSet:GetIntersectingEdges(edgeEndpointA, edgeEndpointB, triangleContainingA)

	local newEdges: { DelaunayTriangleEdge } = {}

	while #intersectedEdges > 0 do
		profilebegin("Clearing intersected edge")
		local currentEdge = table.remove(intersectedEdges) :: DelaunayTriangleEdge -- currentIntersectedTriangleEdge

		-- warn("5.3.3")
		-- 5.3.3: Form quadrilaterals and swap intersected edges
		-- Deduces the data for both triangles
		currentEdge = TriangleSet:FindTriangleThatContainsEdge(currentEdge.EdgeVertexA, currentEdge.EdgeVertexB)
		local intersectedTriangle = TriangleSet:GetDelaunayTriangle(currentEdge.TriangleIndex)
		local oppositeTriangle =
			TriangleSet:GetDelaunayTriangle(intersectedTriangle.AdjacentTriangles[currentEdge.EdgeIndex])

		-- local trianglePoints = triangleSet:GetTrianglePoints(currentEdge.TriangleIndex)
		-- DrawUtil.DrawTriangle(trianglePoints, Color3.new(0, 1, 0))
		-- DrawUtil.DrawTriangle(triangleSet:GetTrianglePoints(intersectedTriangle.AdjacentTriangles[currentEdge.EdgeIndex]), Color3.new(1, 0, 0))

		-- Gets the opposite vertex of adjacent triangle, knowing the fisrt vertex of the shared edge
		local oppositeVertex = NOT_FOUND
		local oppositeSharedEdgeVertex = NOT_FOUND -- The first vertex in the shared edge of the opposite triangle

		local intersectedTriangleEdgePoint = intersectedTriangle.Points[currentEdge.EdgeIndex % 3 + 1]
		for j = 1, 3 do
			-- Comparing with the endpoint B of the edge, since the edge AB is BA in the adjacent triangle
			if oppositeTriangle.Points[j] == intersectedTriangleEdgePoint then
				oppositeVertex = oppositeTriangle.Points[(j + 1) % 3 + 1]
				oppositeSharedEdgeVertex = j
				break
			end
		end

		local nonEdgeVertex = NOT_FOUND
		for j = 1, 3 do
			if
				intersectedTriangle.Points[j] ~= currentEdge.EdgeVertexA
				and intersectedTriangle.Points[j] ~= currentEdge.EdgeVertexB
			then
				nonEdgeVertex = j
				break
			end
		end
		local nonEdgePoint = TriangleSet:GetPointFromIndex(intersectedTriangle.Points[nonEdgeVertex])

		-- DrawUtil.DrawPoint(nonEdgePoint, Color3.new(1, 0, 0), attempts)
		-- DrawUtil.DrawPoint(edgeVertexPointA, Color3.new(0, 1, 0), attempts)
		-- DrawUtil.DrawPoint(edgeVertexPointB, Color3.new(0, 0, 1), attempts)

		local oppositePoint = Points[oppositeVertex]
		-- DrawUtil.DrawPoint(oppositePoint, nil, 2)

		local edgeVertexPointA = Points[currentEdge.EdgeVertexA]
		local edgeVertexPointB = Points[currentEdge.EdgeVertexB]

		-- Check if the quadrilateral is convex
		if DelaunayUtil.IsQuadrilateralConvex(nonEdgePoint, edgeVertexPointA, oppositePoint, edgeVertexPointB) then
			profilebegin("Quadrilateral swap")
			-- Swap edges
			local notInEdgeTriangleVertex = (currentEdge.EdgeIndex + 1) % 3 + 1
			self:_SwapEdges(
				currentEdge.TriangleIndex,
				intersectedTriangle,
				notInEdgeTriangleVertex,
				oppositeTriangle,
				oppositeSharedEdgeVertex
			)

			-- Refreshes triangle data after swapping
			intersectedTriangle = TriangleSet:GetDelaunayTriangle(currentEdge.TriangleIndex)

			-- Check the new diagonal against the intersecting edge
			local newTriangleSharedEdgeVertexA = (currentEdge.EdgeIndex + 1) % 3 + 1
			local newTriangleSharedEdgeVertexB = newTriangleSharedEdgeVertexA % 3 + 1
			local newTriangleSharedEdgeIndexA = intersectedTriangle.Points[newTriangleSharedEdgeVertexA]
			local newTriangleSharedEdgeIndexB = intersectedTriangle.Points[newTriangleSharedEdgeVertexB]

			local newEdge =
				DelaunayTriangleEdge.new(NOT_FOUND, NOT_FOUND, newTriangleSharedEdgeIndexA, newTriangleSharedEdgeIndexB)
			if
				newTriangleSharedEdgeIndexA ~= endpointBIndex
				and newTriangleSharedEdgeIndexB ~= endpointBIndex
				and newTriangleSharedEdgeIndexA ~= endpointAIndex
				and newTriangleSharedEdgeIndexB ~= endpointAIndex
				and DelaunayUtil.AreLineSegmentsIntersecting(
					edgeEndpointA,
					edgeEndpointB,
					TriangleSet:GetPointFromIndex(newTriangleSharedEdgeIndexA),
					TriangleSet:GetPointFromIndex(newTriangleSharedEdgeIndexB)
				)
			then
				--  New triangles edge still intersects with the constrained edge, so it is returned to the list
				table.insert(intersectedEdges, 1, newEdge)
			else
				table.insert(newEdges, newEdge)
			end
			profileend()
		else
			-- Back to the list
			table.insert(intersectedEdges, 1, currentEdge)
		end
		profileend()
	end

	-- 5.3.4. Check Delaunay constraint and swap edges
	for _, edge in newEdges do
		local edgePointA = TriangleSet.Points[edge.EdgeVertexA]
		local edgePointB = TriangleSet.Points[edge.EdgeVertexB]

		if
			(edgePointA == edgeEndpointA and edgePointB == edgeEndpointB)
			or (edgePointB == edgeEndpointA and edgePointA == edgeEndpointB)
		then
			continue
		end

		local currentEdge = TriangleSet:FindTriangleThatContainsEdge(edge.EdgeVertexA, edge.EdgeVertexB)
		local currentTriangle = TriangleSet:GetDelaunayTriangle(currentEdge.TriangleIndex)
		local notSharedVertex = (currentEdge.EdgeIndex + 1) % 3 + 1
		local notSharedPoint = TriangleSet.Points[currentTriangle.Points[notSharedVertex]]
		local oppositeTrianglePoints =
			TriangleSet:GetTrianglePoints(currentTriangle.AdjacentTriangles[currentEdge.EdgeIndex])

		if Triangle2dUtil.IsPointInCircumcircle(notSharedPoint, oppositeTrianglePoints) then
			local oppositeTriangle =
				TriangleSet:GetDelaunayTriangle(currentTriangle.AdjacentTriangles[currentEdge.EdgeIndex])
			local sharedEdgeIndex = self:GetSharedEdge(oppositeTriangle, currentEdge.TriangleIndex)
			self:_SwapEdges(
				currentEdge.TriangleIndex,
				currentTriangle,
				notSharedVertex,
				oppositeTriangle,
				sharedEdgeIndex
			)
		end
	end
end

--[=[
    -- TODO: OPTIMIZE
    Finds the outline of the triangulation excluding the supertriangle.
]=]
function ConstrainedDelaunayTriangulation.GetOutlinePolygon(self: CDTInternal): { Vector2 }
	local outlinePolygon = {}

	local TriangleSet: DelaunayTriangleSet = self.TriangleSet
	local Points = TriangleSet.Points
	local TriangleVertices = TriangleSet.TriangleVertices

	-- Find a triangle that shares a vertex with the supertriangle
	local externalTriangleIndex = NOT_FOUND

	for baseIndex = 0, #TriangleVertices - 1, 3 do
		for j = 1, 3 do
			local vertexIndex = TriangleVertices[baseIndex + j]
			if 1 <= vertexIndex and vertexIndex <= 3 then
				externalTriangleIndex = baseIndex + 1
				break
			end
		end
		if externalTriangleIndex ~= NOT_FOUND then
			break
		end
	end

	if externalTriangleIndex == NOT_FOUND then
		warn("Failed to find external triangle")
		return outlinePolygon
	end

	local currentTriangleIndex = externalTriangleIndex
	repeat
		local triangle = TriangleSet:GetDelaunayTriangle(currentTriangleIndex)
		local outerVertexIndex = NOT_FOUND
		for i, pointIdx in triangle.Points do
			if pointIdx >= 1 and pointIdx <= 3 then
				outerVertexIndex = i
				break
			end
		end

		local nextVertexLocalIndex = if outerVertexIndex == 3 then 1 else outerVertexIndex + 1
		local nextVertexIndex = triangle.Points[nextVertexLocalIndex]
		local previousVertexLocalIndex = if outerVertexIndex >= 2 then outerVertexIndex - 1 else outerVertexIndex + 2
		local previousVertexIndex = triangle.Points[previousVertexLocalIndex]

		-- Check if the next or previous vertex is part of the supertriangle
		if nextVertexIndex >= 1 and nextVertexIndex <= 3 then
			table.insert(outlinePolygon, Points[previousVertexIndex])
			currentTriangleIndex = triangle.AdjacentTriangles[nextVertexLocalIndex]
		elseif previousVertexIndex >= 1 and previousVertexIndex <= 3 then
			table.insert(outlinePolygon, Points[nextVertexIndex])
			currentTriangleIndex = triangle.AdjacentTriangles[previousVertexLocalIndex]
		else
			-- skip triangles with only 1 supertriangle vertex
			table.insert(outlinePolygon, Points[previousVertexIndex])
			currentTriangleIndex = triangle.AdjacentTriangles[outerVertexIndex]
		end
	until currentTriangleIndex == externalTriangleIndex

	-- Debug Visual
	-- for i = 1, #outlinePolygon do
	--     local startPoint = outlinePolygon[i]
	--     local endPoint = outlinePolygon[i % #outlinePolygon + 1]
	--     DrawUtil.DrawLine(startPoint, endPoint)
	-- end

	return outlinePolygon
end

--[=[
    Finds the index of the edge (1 to 3) of a triangle that is shared with another triangle.
    @param triangle -- The triangle whose edge is to be returned.
    @param adjacentTriangleIndex -- The index of the adjacent triangle.
    @return number -- The index of the shared edge in the first triangle, from 1 to 3.
]=]
function ConstrainedDelaunayTriangulation.GetSharedEdge(
	self: CDTInternal,
	triangle: DelaunayTriangle,
	adjacentTriangleIndex: number
): number
	for sharedEdgeVertexLocalIndex = 1, 3 do
		if triangle.AdjacentTriangles[sharedEdgeVertexLocalIndex] == adjacentTriangleIndex then
			return sharedEdgeVertexLocalIndex
		end
	end
	return NO_ADJACENT_TRIANGLE
end

--[=[
    Gets all the triangles that contain any of the vertices of the supertriangle.
    @param outputTriangles -- An array to output the triangles to. If not provided, a new array will be created.
]=]
function ConstrainedDelaunayTriangulation.GetSupertriangleTriangles(self: CDTInternal, _outputTriangles: { number }?)
	local outputTriangles = _outputTriangles or {}
	local TriangleSet: DelaunayTriangleSet = self.TriangleSet
	for pointIndex = 1, 3 do -- Vertices of the supertriangle
		for _, triangleIndex in TriangleSet:GetTrianglesWithVertex(pointIndex) do
			if not table.find(outputTriangles, triangleIndex) then
				table.insert(outputTriangles, triangleIndex)
			end
		end
	end
	return outputTriangles
end

--[=[
    Calculates the bounds of a point cloud, ensuring the minimum position becomes the center of the box.
    @param points {Vector2} -- The points whose bounds are to be calculated.
    @return bounds -- The calculated bounds.
]=]
function ConstrainedDelaunayTriangulation.CalculateBoundsWithLeftBottomCornerAtOrigin(
	self: CDTInternal,
	points: { Vector2 }
): bounds
	profilebegin("CalculateBoundsWithLeftBottomCornerAtOrigin")
	local minX, minY = math.huge, math.huge
	local maxX, maxY = -math.huge, -math.huge

	for _, point in ipairs(points) do
		minX = math.min(minX, point.X)
		minY = math.min(minY, point.Y)
		maxX = math.max(maxX, point.X)
		maxY = math.max(maxY, point.Y)
	end

	local size = Vector2.new(math.abs(maxX - minX), math.abs(maxY - minY))
	local min = Vector2.new(minX, minY)
	local max = Vector2.new(maxX, maxY)
	local center = min + size * 0.5

	local bounds = {
		min = min,
		size = size,
		max = max,
		center = center,
	}

	profileend()
	return bounds
end

--[=[
    Normalizes a list of points according to a bounding box so all of them lay between the 
    coordinates [0,0] and [1,1], while they conserve their relative position with respect to the others.
    @param points {Vector2} The input points to normalize. The points in the list will be updated.
    @param bounds {min: Vector2, size: Vector2} The bounding box in which the normalization is based.
    @return {Vector2} The normalized points.
]=]
function ConstrainedDelaunayTriangulation.NormalizePoints(
	self: CDTInternal,
	points: { Vector2 },
	bounds: bounds
): { Vector2 }
	if DISABLE_NORMALIZATION then
		return points
	end
	profilebegin("NormalizePoints")
	local maxDimension = math.max(bounds.size.X, bounds.size.Y)
	local normalizedPoints = table.create(#points) :: { Vector2 }

	for i, point in points do
		normalizedPoints[i] = (point - bounds.min) / maxDimension
	end
	profileend()
	return normalizedPoints
end

--[=[
    Denormalizes a list of points according to a bounding box so all of them lay between the coordinates determined by the box.
    @param points {Vector2} The points to denormalize. They are expected to be previously normalized.
    @param bounds {min: Vector2, size: Vector2} The bounding box in which the denormalization is based.
    @return {Vector2} The denormalized points.
]=]
function ConstrainedDelaunayTriangulation.DenormalizePoints(
	self: CDTInternal,
	points: { Vector2 },
	bounds: bounds
): { Vector2 }
	if DISABLE_NORMALIZATION then
		return points
	end
	profilebegin("DenormalizePoints")
	local maxDimension = math.max(bounds.size.X, bounds.size.Y)
	local denormalizedPoints = table.create(#points) :: { Vector2 }

	for i, point in points do
		denormalizedPoints[i] = point * maxDimension + bounds.min
	end
	profileend()
	return denormalizedPoints
end

--[=[
    For each triangle, splits its edges into 2 pieces, generating 4 subtriangles. Repeats until no triangle exceeds the maximum area.
    @param maximumTriangleArea number The maximum area all triangles will have after tessellation.
]=]
function ConstrainedDelaunayTriangulation.Tesselate(self: CDTInternal, maximumTriangleArea: number)
	local i = 3 -- Skips supertriangle (indices 1, 2, 3)

	while i <= self.TriangleSet:GetTriangleCount() do
		local triangle = self.TriangleSet:GetDelaunayTriangle(i)
		local isSupertriangle = false

		for _, vertex in ipairs(triangle.Points) do
			if vertex == 1 or vertex == 2 or vertex == 3 then -- Supertriangle vertices
				isSupertriangle = true
				break
			end
		end

		if isSupertriangle then
			i += 1
			continue
		end

		local trianglePoints = self.TriangleSet:GetTrianglePoints(i)
		local triangleArea = Triangle2dUtil.GetArea(trianglePoints)

		if triangleArea > maximumTriangleArea then
			self:_AddPointToTriangulation(trianglePoints[1] + (trianglePoints[2] - trianglePoints[1]) * 0.5)
			self:_AddPointToTriangulation(trianglePoints[2] + (trianglePoints[3] - trianglePoints[2]) * 0.5)
			self:_AddPointToTriangulation(trianglePoints[3] + (trianglePoints[1] - trianglePoints[3]) * 0.5)
			i = 3 -- Restart tessellation
		else
			i += 1
		end
	end
end

return ConstrainedDelaunayTriangulation
